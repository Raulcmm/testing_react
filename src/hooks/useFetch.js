import { useEffect, useRef, useState } from "react";

const useFetch = (api_url) => {
	const isMounted = useRef(true);
	const [state, setState] = useState({ data: null, loading: true, error: null });

	useEffect(() => {
		return () => {
			isMounted.current = false;
		};
	}, []);

	useEffect(() => {
		if (api_url === undefined) return state;
		setState({
			...state,
			loading: true,
		});
		fetch(api_url)
			.then((res) => res.json())

			.then((data) => {
				if (isMounted.current) {
					setState({
						...state,
						data,
						loading: false,
					});
				} else {
					console.warn("Se previnio una fuga");
				}
			})
			.catch((err) =>
				setState({ data: null, loading: false, error: "No se pudo cargar" })
			);
	}, [api_url]);

	return state;
};

export default useFetch;
