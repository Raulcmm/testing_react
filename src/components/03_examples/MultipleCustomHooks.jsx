import React from "react";
import useCounter from "../../hooks/useCounter";
import useFetch from "../../hooks/useFetch";
import "../02_useeffect/useeffect.css";
import Spiner from "../Spiner";

const MultipleCustomHooks = () => {
	const { counter, increment } = useCounter(1);
	const url = `https://www.breakingbadapi.com/api/quotes/${counter}`;
	const { loading, data } = useFetch(url);
	const { author, quote } = !!data && data[0];

	return (
		<div>
			<h1>Custom Hooks</h1>
			<hr />

			{loading ? (
				<Spiner />
			) : (
				<div className="alert alert-success text-center ">
					<blockquote className="blockquote ">
						<p className="mb-5">{quote}</p>
						<footer className="blockquote-footer">{author}</footer>
					</blockquote>
				</div>
			)}
			<button className="btn btn-primary" onClick={increment}>
				Siguiente
			</button>
			<small>{counter}</small>
		</div>
	);
};

export default MultipleCustomHooks;
