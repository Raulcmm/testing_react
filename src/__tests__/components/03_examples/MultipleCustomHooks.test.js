import { shallow } from "enzyme";
import React from "react";

import MultipleCustomHooks from "../../../components/03_examples/MultipleCustomHooks";
// import useCounter from "../../../hooks/useCounter";
import useFetch from "../../../hooks/useFetch";

// //cuando se use en este archivo en lugar de usar el useFetch usa una implementacion del mismo
jest.mock("../../../hooks/useFetch");
// jest.mock("../../../hooks/useCounter");

//SI SE USA UNO MOCK EN ALGUN COMPONENTE ES "OBLIGATORIO " QUE
// SE USE EN TODAS LAS PRUEBAS POR ESO SE PONE EL BEFORE_EACH
// PERO EL FETCH NO POR QUE  SE NECESITAN DIFERENTES DATOS SI FUERAN IGUAL
// SE USA EN EL BEFORE_EACH

describe("Test to MultipleCustomHooks", () => {
	beforeEach(() => {
		// useCounter.mockReturnValue({ counter: 1, increment: () => {} });
		useFetch.mockReturnValue({ data: null, loading: true, error: null });
	});

	test("debe mostrarse correctamente", () => {
		const wrapper = shallow(<MultipleCustomHooks />);
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find("alert").exists()).toBeFalsy();
	});

	test("debe de mostar  informacion", () => {
		useFetch.mockReturnValue({
			data: [{ id: 1, author: "Raul", quote: "Hola mundo" }],
			loading: false,
			error: null,
		});
		const wrapper = shallow(<MultipleCustomHooks />);
		expect(wrapper.find("p").text()).toBe("Hola mundo");
		expect(wrapper.find(".alert").exists()).toBeTruthy();
		expect(wrapper.find("footer").text().trim()).toBe("Raul");
	});

	test("Debe retornar el valor por defecto del counter = 1", () => {
		const wrapper = shallow(<MultipleCustomHooks />);
		expect(wrapper.find("small").text()).toBe("1");
	});

	test("debe regresar el valor de 10 del counter enviado como parametro", () => {
		const wrapper = shallow(<MultipleCustomHooks />);
		expect(wrapper.find("small").text()).toBe("1");
	});

	test("Debe incrementar +1 el counter", () => {
		const wrapper = shallow(<MultipleCustomHooks />);
		wrapper.find("button").simulate("click");
		expect(wrapper.find("small").text()).toBe("2");
	});
});
