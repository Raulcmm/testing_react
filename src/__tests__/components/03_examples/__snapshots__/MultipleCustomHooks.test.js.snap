// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to MultipleCustomHooks debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Custom Hooks
  </h1>
  <hr />
  <Spiner />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente
  </button>
  <small>
    1
  </small>
</div>
`;
